feat: Introducing Tracing with OpenTelemetry API #1537

jimit-j-shah opened the pull request:
* This change introduces Client-Side Tracing using OpenTelemetry API in Datastore Java SDK. 
* Client-Side Tracing using OpenTelemetry in the Datastore v1 Java SDK introduces the ability to trace API calls between Datastore client Datastore Backend by providing a breakdown of time spent in each phase of processing along with vital metadata.

Fixes https://github.com/googleapis/java-datastore/issues/1428 ☕️

jimit-j-shah committed:
feat: Adding TraceUtil interface and its implementation to enable Tracing controls via DatastoreOptions (#1431)

* Adding EnabledTraceUtil, DisabledTraceUtil and TraceUtilTest
* Annotating DatastoreOpenTelemetryOptions to be transient as they're not serializable
* Adding google-auth-library-credentials dependency due to https://github.com/googleapis/java-datastore/actions/runs/8944472794/job/24571458116?pr=1431

jimit-j-shah committed:
feat: Adding Lookup RPC OpenTelemetry Tracing (#1437)

* feat: Adding Lookup RPC OpenTelemetry Tracing
- Removed OpenCensus Tracing
- Added E2E tests with Global and Local OTel SDK
- Moved OTel SDK setup to RemoteDatastoreHelper
- Fixed pom to depend on BOM for all shared dependencies

jimit-j-shah committed:
feat: RunQuery trace instrumentation

jimit-j-shah committed:
Formatting

jimit-j-shah committed:
Formatting

jimit-j-shah committed:
Refactor: s/RUNQUERY/RUN_QUERY

jimit-j-shah committed:
feat: RunAggregationQuery Trace Instrumentation

jimit-j-shah committed:
Build: retiring test assertions for OpenCensus spans - will be replacing this in hermetic integration tests for OpenTelemetry using in-memory span exports (in addition to ITE2ETraceTest.java).

jimit-j-shah committed:
Formatting

jimit-j-shah committed:
feat: Adding Commit RPC Trace Instrumentation (#1440)

- Added end-to-end test for Datastore operationsput, add, update and delete.
- Updated E2E Test to use the namespace correctly for efficient clean-up of test data

jimit-j-shah committed:
Fixing @Test annotation missed after merge

jimit-j-shah committed:
Formatting

jimit-j-shah committed:
feat: Add Transaction tracing
test: transactionalLookupTest

jimit-j-shah committed:
test: Transaction test for RunInTransaction - need to fix trace instrumentation for RunIn..

jimit-j-shah committed:
feat: RunQuery trace instrumentation (#1441)

* feat: RunQuery trace instrumentation

jimit-j-shah committed:
feat: RunAggregationQuery instrumentation (#1447)

* feat: RunQuery trace instrumentation

* Formatting

* Formatting

* Refactor: s/RUNQUERY/RUN_QUERY

* feat: RunAggregationQuery Trace Instrumentation

* Build: retiring test assertions for OpenCensus spans - will be replacing this in hermetic integration tests for OpenTelemetry using in-memory span exports (in addition to ITE2ETraceTest.java).

* Formatting

* Fixing @Test annotation missed after merge

* Formatting

jimit-j-shah committed:
Adding transaction span names

jimit-j-shah committed:
TransactionLookupTest

jimit-j-shah committed:
feat: support for transactional operations
- tested using newTransaction() and runInTransaction()

jimit-j-shah committed:
Revert "feat: support for transactional operations"

This reverts commit 10341c0b97cbc2025f9f928ce8cb09d5c036a5b3.

jimit-j-shah committed:
feat: support for transactional operations (#1468)

* feat: support for transactional operations
- tested using newTransaction() and runInTransaction()

jimit-j-shah committed:
feat: Allocateid tracing (#1488)

* feat: Adding tracing for AllocateIds RPC

* formatting

* 🦉 Updates from OwlBot post-processor

See https://github.com/googleapis/repo-automation-bots/blob/main/packages/owl-bot/README.md

---------

Co-authored-by: Owl Bot <gcf-owl-bot[bot]@users.noreply.github.com>

jimit-j-shah committed:
feat: Add tracing for ReserveIds operation (#1490)

- added end-to-end test

jimit-j-shah committed:
fix: Fixed Span nesting for `ReadWriteTransactionCallable` by using parent SpanContext instead of just parent Context (#1495)

* fix: Fixed the TraceUtil.startSpan method to use `SpanContext` for linking with the parent instead of `Context`.
- This fixes the hierarchy of Spans appearing in a transaction under a Run method.
- Tested using existing transaction test

* Fixed commit reordering and typos

* fix: lint errors

* fix: Refactored the ReadWriteTransactioncallable.call method to use startSpan idiomatically
- TraceUtil.startSpan needs more debugging
- return DefaultTracerProvider instance (no-op) when initializing DisabledTraceUtil - this fixes the unit tests in DatastoreTest.testRunInTransactionWithReadWriteOption

* feat: Added tracing for Transaction.RunQuery (#1499)

* feat: Added span for Transactional RunQuery
- tested

* fix: lint

* fix: patch apply issues

* fix: refactor using boolean flag

* fix: s/startSpan/startSpanWithParentContext

jimit-j-shah committed:
test: Additional Transaction Testing and cleanup OpenCensus usage (#1505)

* test: newTransactionReadWriteTraceTest

* fix: test literal

* feat: Added tests for transaction cases

* fix: Cleanup OpenCensus dead code

* fix: updating version from 2.20.1 -> 2.21.0

* fix: reverting version from 2.21.0 -> 2.20.1

* fix: Adding an exception to the clirr-maven-plugin for an internal API parameter change from com.google.cloud.datastore.TraceUtil -> com.google.cloud.datastore.telemetry.TraceUtil

* fix: Fixing the differenceType in clirr exception

* fix: add an exception for removal of an internal class (com.google.cloud.datastore.TraceUtil)

* fix: fixing incomplete difference details for type 7005

* fix: Fixing `to` of the difference to be the entire signature

* fix: typo

jimit-j-shah committed:
test: Adding ITTracingTest to verify events and span attributes (whic… (#1514)

* test: Adding ITTracingTest to verify events and span attributes (which are not verified in ITE2ETracingTest) due to TraceClient API limitations.
- This test uses InMemorySpanExporter to read the generated Otel span data by the test process to verify generated span data as it were before exporting to a backend. None of the span data is exported to a durable backend.
- This test is still an E2E test as it requires a project to send RPCs to.

* fix: fixing compilation error due to missing pom dependency.

* test: Test for AllocateId and ReserveId rpcs

* test: Commit/Put/Update/Delete tests

* test: Added fixes and test for RunQuery event

jimit-j-shah committed:
test: Additional Transaction tests and AggregationQuery test (#1518)

* test: ReadWrite Transaction test

* test: Added test for Transactional RunQuery and Transaction Rollback

* test: runInTransaction API tracing test
- Fixed setting of common span attributes to spans in runInTransaction
- Removed some gRPC related channel attributes that are not present in this Datastore version, yet.

jimit-j-shah committed:
fix: Undelete gRPC upgrade docs

jimit-j-shah committed:
fix: Undo merge mistakes

jimit-j-shah committed:
fix: Updating span event strings (#1539)

* fix: Fixing user-facing span names in line with go/firestore-client-trace-catalog

* fix: updating bom dependency version to fix https://github.com/googleapis/java-datastore/actions/runs/10256441634/job/28375496112?pr=1539

jimit-j-shah committed:
Fix: typo in test causing integration test failure (#1556)

https://btx.cloud.google.com/invocations/c11a2e8b-4494-4ddc-a77e-cf2bcbcf5254/targets/cloud-devrel%2Fclient-libraries%2Fjava%2Fjava-datastore%2Fpresubmit%2Fintegration;config=default/log

jimit-j-shah committed:
fix: opentelemetry-sdk should only be used as a Test Dependency

jimit-j-shah committed:
fix: Update opentelemetry.version - this also fixes the tests failing in https://github.com/googleapis/java-datastore/actions/runs/10891578591/job/30222786908

jimit-j-shah committed:
fix: Replacing attribute key values w/ constants

jimit-j-shah committed:
fix: opentelemetry.version to fix RequireUpperBoundDeps check https://github.com/googleapis/java-datastore/actions/runs/10892403348/job/30225154043?pr=1576

cindy-peng commented on google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:
```diff
@@ -125,20 +154,57 @@ void setPrevTransactionId(ByteString transactionId) {
     public T call() throws DatastoreException {
-      transaction = datastore.newTransaction(options);
-      try {
+      // TODO Instead of using OTel Spans directly, TraceUtil.Span should be used here. However,
```
Nit: is it possible to link the bug ID here?

jimit-j-shah replied:
This has been fixed in the latest PR so this comment is oboslete.

cindy-peng commented on google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:
```diff
@@ -99,12 +122,18 @@ static class ReadWriteTransactionCallable<T> implements Callable<T> {
     private volatile TransactionOptions options;
     private volatile Transaction transaction;
 
+    private final com.google.cloud.datastore.telemetry.TraceUtil.SpanContext parentSpanContext;
```
Can we import these as headers they appear several times in this file?

jimit-j-shah replied:
Not using this anymore.

cindy-peng commented on google-cloud-datastore/src/main/java/com/google/cloud/datastore/DatastoreImpl.java:
```diff
@@ -125,20 +154,57 @@ void setPrevTransactionId(ByteString transactionId) {
+      SpanBuilder spanBuilder =
+          otelTraceUtil
+              .getTracer()
+              .spanBuilder(com.google.cloud.datastore.telemetry.TraceUtil.SPAN_NAME_TRANSACTION_RUN)
```
Same as above, considering import as headers

jimit-j-shah replied:
Done.

cindy-peng commented on google-cloud-datastore/src/main/java/com/google/cloud/datastore/spi/v1/HttpDatastoreRpc.java:
```diff
@@ -40,6 +39,7 @@
 import com.google.datastore.v1.RunAggregationQueryResponse;
 import com.google.datastore.v1.RunQueryRequest;
 import com.google.datastore.v1.RunQueryResponse;
+import io.opencensus.trace.Tracing;
```
Do we need to add this?

jimit-j-shah replied:
Due to removal of the datastore.TraceUtil class, this needs to be added to continue supporting opencensus tracing for HTTP.

jimit-j-shah committed:
fix: Create Span hierarchy using parent Span (#1580)

* fix: Replace use of TraceUtil.SpanContext w/ TraceUtil.Context

* fix: Fixing how span hierarchy is created across threads - using Span instead of Context

* fix: cleaning up startSpan(spanName, parentContext) variant

* fix: add TracedReadWriteTransactionCallable to bifurcate tracing enabled/disabled paths for the Transaction callback.
- This change implements the idiomatic way to express nested spans as described in https://opentelemetry.io/docs/languages/java/instrumentation/#create-nested-spans

* fix: cleanup

* fix: cleanup

* fix: cleanup

jimit-j-shah committed:
fix: formatting and import refactoring

jimit-j-shah committed:
Merge branch 'main' into tracing-main-4

cloud-java-bot committed:
chore: generate libraries at Thu Sep 19 18:35:54 UTC 2024

ehsannas reviewed (approved):
LGTM. Thanks @jimit-j-shah . I haven't done an in-depth review since I'm not a Datastore expert, but I've read through the public APIs and the tracing design and they look great.
